### PR TITLE
[AutoDiff] Fix unexpected nondifferentiability error.

### PR DESCRIPTION
**Pull request by dan-zheng**
Fix unexpected error when differentiating protocol requirements and class
methods with multiple `@differentiable` attributes.

Various gardening included.

Resolves [TF-676](https://bugs.swift.org/browse/TF-676).

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -1960,6 +1960,43 @@ static SubstitutionMap getSubstitutionMap(
   return substMap;
 }
 
+/// Finds the `[differentiable]` attribute on the specified original function
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Todo: fix doc comment and rename function to `lookUpMinimalDifferentiableAttrAndIndexSubset`.

File: test/AutoDiff/autodiff_diagnostics.swift [hunk 2]
```diff
@@ -170,6 +170,33 @@ struct TF_305 : Differentiable {
...
+}
+
+// TF-676: Test differentiation of class method with multiple `@differentiable`
+// attributes.
+class MultipleDiffAttrsClass : Differentiable {
+  @differentiable(wrt: (self, x))
+  @differentiable(wrt: x)
+  func f(_ x: Float) -> Float { x }
+}
+func testMultipleDiffAttrsClass<C: MultipleDiffAttrsClass>(_ c: C, _ x: Float) {
```
Comment (rxwei) [inline test/AutoDiff/autodiff_diagnostics.swift]: what is it upcasting to?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -5375,6 +5398,9 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
   NO_ADJOINT(StrongRetainUnowned)
   NO_ADJOINT(DestroyValue)
   NO_ADJOINT(DestroyAddr)
+
+  // Class-related instructions.
+  NO_ADJOINT(ClassMethod) // Active only because it takes a class operand.
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: How can it be active? The result is a function, and function will only be made active when we need their context cotangents.

File: test/AutoDiff/autodiff_diagnostics.swift (hunk 2 above)
Comment (dan-zheng) [reply test/AutoDiff/autodiff_diagnostics.swift]: Upcasting `c` to `MultipleDiffAttrsClass`.
```
[AD] Unhandled instruction in PullbackEmitter:   %5 = upcast %0 : $C to $MultipleDiffAttrsClass  // users: %8, %7, %6
[AD] Diagnosing non-differentiability.
[AD] For instruction:
  %5 = upcast %0 : $C to $MultipleDiffAttrsClass  // users: %8, %7, %6
[AD] With invoker:
(differentiation_invoker autodiff_function_inst=(  %12 = autodiff_function [wrt 0 1] [order 1] %11 : $@callee_guaranteed (@guaranteed C, Float) -> Float // users: %80, %32, %14, %13
))
tf-676.swift:21:26: error: function is not differentiable
  _ = gradient(at: c, x) { c, x in c.f(x) }
                         ^~~~~~~~~~~~~~~~~~
tf-676.swift:21:36: note: expression is not differentiable
  _ = gradient(at: c, x) { c, x in c.f(x) }
                                   ^
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Oops, that was a total blunder, thanks.
Removed in https://github.com/apple/swift/pull/26313/commits/7aa294508721bfcc2437a157ea46e2df6e1d9dd0.

File: test/AutoDiff/autodiff_diagnostics.swift (hunk 2 above)
Comment (rxwei) [reply test/AutoDiff/autodiff_diagnostics.swift]: Oh ok. Let's fix that later.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -965,41 +965,77 @@ class ADContext {
...
+      //
+      // NOTE(TF-642): `da` may come from a un-partial-applied function and
+      // have larger capacity than the desired indices. We expect this logic to
+      // go away when `partial_apply` supports `@differentiable` callees.
+      if (daIndexSet->isSupersetOf(indexSet->extendingCapacity(getASTContext(),
+              daIndexSet->getCapacity())) &&
+          // fewer parameters than before
+          (minimalIndexSet->isEmpty() ||
+           daIndexSet->getNumIndices() < minimalIndexSet->getNumIndices())) {
+            minimalAttr = da;
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Fix indentation.

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks, done in https://github.com/apple/swift/pull/26313/commits/1e535fa27077fb836653365b953b7c564e5dbf58.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**